code/modules/mob/living/simple_animal/bot/turtle/turtle_cartridge.dm

```diff
@@ -0,0 +1,280 @@
+
+	var/wait = 1
+
+	if (stack.len)
+		var/datum/stack_item/I = stack[stack.len]
+		if (command == "END")
+			if (istype(I, /datum/stack_item/cmd_while))
+				if(I.skipping)
+					stack.len --
+				else
+					if(process_expression(I.condition))
+						line = I.line
+					else
+						stack.len --
+			else if(istype(I, /datum/stack_item/cmd_if))
+				stack.len --
+			else
+				stack.len --
+
+		else if (I.skipping)
+			if(command == "IF" || command == "WHILE") //acount for if and while even if we're skipping so we match up ends
+				var/datum/stack_item/NI = new
+				NI.skipping = TRUE
+				stack += NI
+
+			line++
+			if (line > lines.len)
+				stop()
+			return
+
+	else if(command == "END")
+		throw_error("MISMATCHED END")
+
+	if(!running) return
+
+	switch(command)
+		if("MOVE", "GOTO", "WAIT")
+			if(arguments.len < 1) throw_error("INVALID ARG AMOUNT", arguments.len)
+			process_val(process_expression(arguments))
+		if("SET")
+			if(arguments.len < 2) throw_error("INVALID ARG AMOUNT", arguments.len)
+			if(!is_varname(arguments[1])) throw_error("INVALID VAR NAME", arguments[1])
+			if(!(arguments[1] in variables) && variables.len == memory_limit) throw_error("OUT OF MEMORY")
+			process_expression(arguments.Copy(2))
+		if("WHILE", "IF")
+			if(arguments.len < 1) throw_error("CONDITION NOT SUPPLIED")
+			process_expression(arguments)
+
+	if(!running) return //dont run any of this if we have stopped the program due to an error
+
+	switch(command)
+		if("SAY")
+			if(full_input) turtle?.speak(format_string(full_input))
+			wait = 10
+		if("BEEP")
+			turtle.beep("ping")
+			wait = 10
+		if("MOVE")
+			turtle?.Move(get_step(get_turf(turtle), process_expression(arguments)))
+			wait = 2
+		if("GOTO")
+			line = process_expression(arguments)
+		if("WAIT")
+			wait = process_expression(arguments)
+		if("SET")
+			variables[arguments[1]] = process_expression(arguments.Copy(2))
+		if("STOP")
+			stop()
+		if("IF")
+			var/datum/stack_item/cmd_if/I = new
+			I.condition = arguments
+			I.line = line
+			I.skipping = !process_expression(arguments)
+			stack += I
+		if("WHILE")
+			var/datum/stack_item/cmd_while/I = new
+			I.condition = arguments
+			I.line = line
+			I.skipping = !process_expression(arguments)
+			stack += I
+
+		if("#")
+			wait = 0 //basically ignore this line since it's a comment
+		if("END")
+		else
+			return throw_error("INVALID COMMAND", command)
+
+	if(wait) sleep(wait)
+	if(command != "GOTO")
+		line++
+	if (line > lines.len)
+		stop()
+
+/obj/item/disk/turtle_cartridge/test
+	name = "TurtleOS v1.2 Test Cartridge"
+	lines = list(
+		"WAIT 20",
+		"BEEP",
+		"SET n 0",
+		"SAY Hello World! \[n\]",
+		"BEEP",
+		"SET n n ADD 1",
+		"SET n n MUL n",
+		"MOVE east",
+		"MOVE north",
+		"GOTO 4"
+	)
+
+/obj/item/disk/turtle_cartridge/test2
+	name = "TurtleOS v1.2 Test Cartridge"
+	lines = list(
+		"SET n 0",
+		"WHILE n LT 20",
+		"IF n MOD 2",
+		"SAY \[n\] IS ODD",
+		"IF n EQ 11",
+		"SAY \[n\] IS ODD and 10",
```

Comment:
code doesn't match the say

Reply:
huh?

Reply:
oh thats just a test program no need to review it

Reply:
All shall be reviewed.